### PR TITLE
[Not for 0.15] adding General Bayesian Tests

### PR DESCRIPTION
**Pull request by FBartos**
analysis that allows specifying different prior distributions for a parameter and test between using different data likelihoods

### CODE REVIEW

File: inst/help/summarystatsgeneralbayesiantests.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+### Input
+---------
+
+#### Data
+The input data needs to contain the following elements:
+
+- Likelihood specifying the distribution of the data
+  - Normal with mean and standard deviation
+  - Student-t with mean, standard deviation, and degrees of freedom
```
Comment (Kucharssim) [inline inst/help/summarystatsgeneralbayesiantests.md:15]:
```suggestion
  - Student-t with location, scale, and degrees of freedom
```
I think it's more common to call these parameters that way?

File: inst/help/summarystatsgeneralbayesiantests.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+#### Alternative hypotheses
+The specification of posibly multiple alternative hypotheses for the parameter.
+
+
+#### Plots
+  - Priors: Plot parameter prior distributions.
+  - Prior predictions: Plot prior predictive distributions.
+  - Likelihood: Plot data likelihood.
+  - Posteriors: Plot parameter posterior distributions.
+    - Priors: Add prior distribution to the posterior distribution plots.
```
Comment (Kucharssim) [inline inst/help/summarystatsgeneralbayesiantests.md:34]:
```suggestion
  - Priors: Add prior distribution to the posterior distribution plots. 
```

**Review by Kucharssim — CHANGES_REQUESTED** [review]
The code looks ok, I left some minor comments.

However, the analysis does not run for me, specifically it just throws unexpected error upon opening it and no amount of fiddling with the options changes that. Once you fix that, I will try it again.

rename to .jasp to see the error
[generalTestError.zip](https://github.com/jasp-stats/jaspSummaryStatistics/files/7225553/generalTestError.zip)

File: inst/Description.qml [hunk 3]
```diff
@@ -73,4 +73,16 @@ Description
         title:	qsTr("S.S. Bayesian A/B Test")
         func:	"SummaryStatsABTestBayesian"
     }
+
+	GroupTitle
+	{
+		title: qsTr("Miscellaneous")
+		icon: "analysis-bayesian-crosstabs.svg"
+	}
```
[the remark this answers is not in the document]
Comment (FBartos) [reply inst/Description.qml:81]: This was a name we settled on with EJ at the end. Any second thoughts @EJWagenmakers ?

**Comment by FBartos** [comment]
Oh, I see that there were significant updates to the package since I submitted the PR.

**Comment by FBartos** [comment]
so, that should be fixed & I also change themeJasp to themeJaspRaw as suggested by Don elsewhere

File: inst/qml/SummaryStatsGeneralBayesianTests.qml [hunk 4]
```diff
@@ -0,0 +1,622 @@
...
+// License along with this program.  If not, see
+// <http://www.gnu.org/licenses/>.
+//
+
+
+import QtQuick 2.8
+import QtQuick.Layouts 1.3
+import JASP.Controls 1.0
+import JASP.Widgets 1.0
+
```
Comment (Kucharssim) [inline inst/qml/SummaryStatsGeneralBayesianTests.qml:24]:
```suggestion
import JASP 1.0
```

without this, the qml does not know how to use `JASP.None` to disallow inputing 0 for positive only parameters

**Review by Kucharssim — CHANGES_REQUESTED** [review]
Now it works!

I left one comment in the code review.

Some additional issues:

1) The lines in the prior and posterior plots appear unnecessarily thin. I would suggest making them thicker
Screenshot 2021-09-27 at 12 31 30

2) Would be really cool to add tables of the summary statistics of the posteriors. Not for this release of course, but I hope you'll add that at some point!

3) I am confused by the likelihood section somewhat. Intuitively, it seems to me that the analysis should take the sufficient statistics for the selected likelihood function to compute the results. This is for example the case when I select `Binomial` likelihood. However, for `Normal`, I can only input `mean` and `sd`, but not `sample size`, but you need that to evaluate the likelihood function from the summary statistics. So I am not sure what the analysis does without specifying the sample size: is it assuming you are observing a single trial with uncertainty or something? Or is `sd` actually `se` of the mean? Similarly, I think the analysis is a little bit cryptic about what it actually tests. Again, with binomial, it's clear that you test the parameter theta, but with normal, for example, it's a bit unclear that the only thing you can test is the mean, and it also does not communicate what it does with the variance parameter under each individual model.

**Comment by FBartos** [comment]
1) Hmm, the figures are generated with the ```bayesplay``` package. Is there an easy way of editing the line thickness of the generated ggplots?
2) Good idea, I added an issue for this (https://github.com/jasp-stats/INTERNAL-jasp/issues/1561; not sure that it will be feasible with the package since it uses numerical integration though).
3) Agree, it's not really clear and the package documentation does not help (```?bayesplay::likelihood```). Not really sure what to do about that.

**Comment by vandenman** [comment]
Looking at `bayesplay:::make_likelihood.normal`, the analysis treats the normal as a single observation. You'd have to enter the standard error yourself to get the distribution of sufficient statistics. This is probably not ideal.

**Comment by Kucharssim** [comment]
> 1. Hmm, the figures are generated with the bayesplay package. Is there an easy way of editing the line thickness of the generated ggplots?

I think you could use `scale_size_manual()`, but not sure whether that will work.

> 3. Agree, it's not really clear and the package documentation does not help (?bayesplay::likelihood). Not really sure what to do about that.

> Looking at bayesplay:::make_likelihood.normal, the analysis treats the normal as a single observation. You'd have to enter the standard error yourself to get the distribution of sufficient statistics. This is probably not ideal.

Right. Yeah, according to the website, you need to put in SE manually in order it to work properly (https://bayesplay.github.io/bayesplay/articles/basic.html). I think we want to avoid the situation that people will add their standard deviation into the analysis thinking that the analysis will test the mean properly, as they would have to calculate the SE instead. Should we rename the option then? It really depends on whether you consider it as a data from a single uncertain observation (where the SD would be appropriate), or whether it's a summary statistic of many observations (in which SE would be appropriate). Also, the analysis assumes then that the variance parameter is known I think, so I guess it's good to at least communicate it clearly with a footnote or something.

**Comment by FBartos** [comment]
So, the ```scale_size_manual()``` didn't work, so I had to extract the plotting functions from the package. The line thickness should be in line with the JASP theme.

And I changed the SD label to SE to make it clearer.

**Comment by AlexanderLyNL** [comment]
> Looking at `bayesplay:::make_likelihood.normal`, the analysis treats the normal as a single observation. You'd have to enter the standard error yourself to get the distribution of sufficient statistics. This is probably not ideal.

Perhaps mention this to the package maintainer.

**Comment by FBartos** [comment]
OK, mentioned.

**Review by Kucharssim — APPROVED** [review]
> I needed to extract all plotting functions from the bayesplay package for it. I guess it's gonna bite us if/when the author updates the internal functions...

Yeah, that's unfortunate. I do think that the plots look much better now though. I think the ideal scenario would be to ask the maintainer of the package to allow users to specify the width of the lines - are you in touch with them so that you can ask if that's possible? Alternatively, you can submit a feature request on their GitHub page. If that's implemented, we can revert back to using the plotting functions from the package. I also give you a bit of a leeway with the snake case that is in the functions that you copied over.

**Pull request merged**